environment/src/main/java/com/sequenceiq/environment/network/service/SubnetIdProvider.java

```diff
@@ -1,27 +1,58 @@
 package com.sequenceiq.environment.network.service;
 
+import static com.sequenceiq.cloudbreak.common.mappable.CloudPlatform.AZURE;
+
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Optional;
 
+import org.apache.commons.lang3.StringUtils;
 import org.springframework.stereotype.Component;
 
 import com.sequenceiq.cloudbreak.cloud.model.CloudSubnet;
+import com.sequenceiq.cloudbreak.common.mappable.CloudPlatform;
 import com.sequenceiq.common.api.type.Tunnel;
 import com.sequenceiq.environment.network.dto.NetworkDto;
 
 @Component
 public class SubnetIdProvider {
 
-    public String provide(NetworkDto network, Tunnel tunnel) {
-        if (network != null && network.getSubnetIds() != null && !network.getSubnetIds().isEmpty()) {
-            return getSubnetIdByPreferedSubnetType(network, tunnel.useCcm());
-        } else {
-            return null;
+    public String provide(NetworkDto network, Tunnel tunnel, CloudPlatform cloudPlatform) {
+        return network != null && network.getSubnetIds() != null && !network.getSubnetIds().isEmpty() && !network.getSubnetMetas().isEmpty()
+                ? getSubnetIdByPreferedSubnetType(network, tunnel.useCcm(), cloudPlatform)
+                : null;
+    }
+
+    private String getSubnetIdByPreferedSubnetType(NetworkDto network, boolean preferPrivate, CloudPlatform cloudPlatform) {
+        List<CloudSubnet> subnetMetas = new ArrayList<>(network.getSubnetMetas().values());
+        return AZURE.equals(cloudPlatform)
+                ? subnetMetas.get(0).getId()
+                : getSubnetIdByPreferedSubnetTypeAws(subnetMetas, preferPrivate);
+    }
+
+    private String getSubnetIdByPreferedSubnetTypeAws(List<CloudSubnet> subnetMetas, boolean preferPrivate) {
```

Comment:
I'm missing the scenario here, when we prefer private, but we don't have one, we should choose a public with mapPublicIpOnLaunch enabled if it's possible, and fall back to any subnet if it fails

Reply:
Yes, that is also useful, added it. Thanks.